crates/compiler/parse/src/src64.rs

```diff
@@ -0,0 +1,437 @@
+            }
+            None => Err(FileErr::FileWasTooBig(file_size)),
+        }
+    }
+}
+
+fn round_up_to_nearest_64(num: usize) -> usize {
+    // Round up to the nearest 64. (Writing this out as 64 - 1 so it's clearer where the numbers came from.)
+    // We can do saturating addition here rather than overflow checking, because if we overflow usize::MAX,
+    // we will most definitely be over the max source file size and return None anyway.
+    (num.saturating_add(64 - 1)) & !(64 - 1)
+}
+
+/// Safety: capacity must be a multiple of 64, and must be at least 64.
+unsafe fn allocate_and_pad_with_newlines(
+    arena: &Bump,
+    capacity: usize,
+) -> Option<&mut [MaybeUninit<u8>]> {
+    // Compare capacity here instead of size because this file limit is based on what we can record row and line
+    // numbers for, and those can theoretically oveflow on the trailing newlines we may have added.
+    // This distinction will most likely come up in practice zero times ever, but it could come up in fuzzing.
+    if capacity > MAX_ROC_SOURCE_FILE_SIZE {
+        return None;
+    }
+
+    debug_assert!(capacity >= 64);
+    debug_assert!(capacity % 64 == 0);
+
+    // Safety: the rules we follow are https://doc.rust-lang.org/core/alloc/struct.Layout.html#method.from_size_align_unchecked
+    // `align` is valid because it's hardcoded, and we already rounded `capacity` up to something even bigger.
+    let layout = unsafe { Layout::from_size_align_unchecked(capacity, 16) };
+
+    // We have to use alloc_layout here because we have stricter alignment requirements than normal slices.
+    let buf_ptr: NonNull<u8> = arena.alloc_layout(layout);
+
+    // Safety: we just allocated this pointer to have a capacity of `size` bytes.
+    let contents = unsafe { slice::from_raw_parts_mut(buf_ptr.as_ptr() as _, capacity) };
+
+    // Safety: `contents` has a length of `capacity`, which has been rounded up to a multiple of 64.
+    unsafe { fill_last_64_bytes_with_newlines(contents) };
+
+    Some(contents)
+}
+
+/// This is branchless so there can't be mispredictions. We know the slice's length is a multiple of 64,
+/// so we can just always do four SIMD writes and call it a day. (Eight if we don't have SIMD.)
+///
+/// Safety: this slice must have an alignment of at least 64,
+/// and its length must be both at least 64 and also a multiple of 64.
+unsafe fn fill_last_64_bytes_with_newlines(slice: &mut [MaybeUninit<u8>]) {
+    let len = slice.len();
+
+    debug_assert_eq!(
+        slice.as_ptr() as usize % 16,
+        0,
+        "The slice's alignment must be at least 16."
+    );
+    debug_assert_eq!(len % 64, 0, "The slice's length must be a multiple of 64.");
+    debug_assert!(len >= 64, "The slice's length must be at least 64.");
+
+    // Safety: this function's docs note that it must be given a slice with at least 64 bytes in it.
+    let last_64_bytes = slice.as_mut_ptr().add(len - 64);
+
+    #[cfg(target_feature = "sse2")]
+    {
```

Comment:
should there not be a fallback for if you have none of these features?

Reply:
We do! 😃 

It's right after the neon `cfg` block.